### PR TITLE
(PPS-107): Feat/single table driver

### PR DESCRIPTION
**Pull request by BinamB**
Jira Ticket: [PXP-xxxx](https://ctds-planx.atlassian.net/browse/PXP-xxxx)
- [ ] Remove this line if you've changed the title to (PXP-xxxx): <title>

Coverage went down by 2% because we removed `test_setup.py`. The coverage went down for the `utils.py` file that contains the old, deprecated method of database migration that uses `sqlite3`.

### New Features
- Implemented single table architecture to IndexD with a single new `record` table (Configurable and requires migration)

### Improvements
- Removed `sqlite` from unit tests. Unit tests now only uses `postgres`

### Dependency updates
- `sqlalchemy` ~1.3.3 -> ^1.4.0

### Deployment changes

- sqlite databases are not supported anymore
- To run single table migration check this doc: https://github.com/uc-cdis/indexd/blob/master/docs/migration_to_single_table_indexd.md

### CODE REVIEW

**Review by Avantol13 — CHANGES_REQUESTED** [review]
this isn't a full review but figured I share these comments I had so far

**Review by Avantol13 — CHANGES_REQUESTED** [review]
I had some comments before I left for PTO that I forgot to send, apologies. Here they are now

File: indexd/default_settings.py
Comment (BinamB) [reply indexd/default_settings.py]: Yeah sqlite doesn't handle some of the newer data types. I moved all the test to use postgres instead of sqlite too, just to stay consistent. I saw there was a mix.

File: indexd/index/drivers/single_table_alchemy.py
Comment (Avantol13) [inline indexd/index/drivers/single_table_alchemy.py:1239]: we could add a simple caching strategy here where we store some global and clear it if records get added. Maybe out of scope for this but worth considering. Same thing for num records

File: migrations/versions/bb3d7586a096_createsingletable.py
Comment (Avantol13) [inline migrations/versions/bb3d7586a096_createsingletable.py:20]: I think we need another PR after this gets merged to actually clean up the old tables. We can't forget or someone in the future will be super confused.

File: tests/conftest.py
Comment (Avantol13) [inline tests/conftest.py:23]: isn't there a configuration value for this that the tests use?

File: tests/conftest.py
Comment (Avantol13) [inline tests/conftest.py:26]: quick docstring pls

File: tests/conftest.py
Comment (Avantol13) [inline tests/conftest.py:69]: quick docstring pls

File: tests/postgres/migrations/test_bb3d7586a096_createsingletable.py
Comment (Avantol13) [inline tests/postgres/migrations/test_bb3d7586a096_createsingletable.py:4]: quick docstring pls

File: tests/postgres/migrations/test_bb3d7586a096_createsingletable.py
Comment (Avantol13) [inline tests/postgres/migrations/test_bb3d7586a096_createsingletable.py:43]: quick docstring pls

**Review by Avantol13 — CHANGES_REQUESTED** [review]

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:2]: Can you add a quick description here with how to use this?

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:265]: dead code, remove if not using

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:71]: I imagine this batch_size could be much larger. Where's the job that will run this? How much RAM do you plan on providing? Indexd records text is pretty darn small. Maybe 1000 characters on average? 1000 chars / record, 1 char = 1 byte in ASCII, 1kb per record. so 1000 records = 1MB. If we have 1 GBs of usable RAM, we could probably go up to like 10000 records.

Comment (Avantol13) [reply bin/migrate_to_single_table.py:71]: I'm not sure we _should_ do that, but if we can reliably, it might be faster (reducing the total number of queries and commits to the db). Would outputting to a file and using /COPY be faster? Don't you guys use /COPY for something already? Did you consider this?

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:148]: lets get some docstrings in these functions

File: indexd/index/drivers/single_table_alchemy.py
Comment (BinamB) [reply indexd/index/drivers/single_table_alchemy.py:55]: removing it, I cant remember why i did that. Its been a while.

File: bin/migrate_to_single_table.py
Comment (BinamB) [reply bin/migrate_to_single_table.py:71]: I did consider this. I think there's a couple of things here:
1. We had a conversation about whether to use sql+bash script vs python. We went with python because it would be easier for any gen3 user to use it. 
2. should we do python + /copy? My concern with this is that this would blow up for something like dcf where we have about 56mil records. I think we'd use up a lot of memory. We could do it in batches though 🤔

File: migrations/versions/bb3d7586a096_createsingletable.py
Comment (BinamB) [reply migrations/versions/bb3d7586a096_createsingletable.py:20]: Yeah definitely. This migration can be an in-between migration where both options are available.

File: bin/migrate_to_single_table.py
Comment (Avantol13) [reply bin/migrate_to_single_table.py:71]: Yeah, it's been a while since we've had the design conversations around the options here. Python makes sense, - we're already so close with this. I was just thinking out loud and not remembering previous conversations. Let's not shift gears at this point. I am curious if you were able to tune these numbers and avoid using OFFSET and if that helped

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:52]: minor thing: I think it's more intuitive if it's `None` instead of `False`

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:78]: use logger/logging instead of pure prints

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:149]: SQLAlchemy lists this as legacy https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.bulk_save_objects for 2.0 and beyond. I think since this is a one time run and we're not ported to 2 yet, we're okay

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:158]: instead of this, you should have an except block for the specific Error type that raises that, you can have multiple except blocks in Python so you can split this. out and keep the general rollback case on Exception too

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:166]: Please what ? I think the second line didn't get committed

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:157]: What's the intended behavior here? It looks like you rollback this batch but then just continue anyway, but you don't have a log for last seen guid. I think you need to do that before exiting. Also I think this should just retry the batch a few times before giving up. 

In fact, this code is pretty nested and getting hard to read. I'd break things out into functions and you can have a process_batch function that you use exponential backoff on

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:248]: are you intending to not reraise an error here? so we just skip that field and it gets set as None? I don't think we want that behavior b/c then some records could get "successfully" submitted but if one of these functions failed, that field would be None. This is a good reason to try and avoid early returns b/c that behavior is not obvious. But if you had a guaranteed return value at the end, it would be.

e.g.

```
field_value = None
try:
  get
except: 
  whoops
return field_value
```

That being said, I don't think we want that behavior. I think we want to raise the exception higher and then exponentially backoff trying 1 or 2 more times to get the info for the record.

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:304]: do you know how this happens in the first place?

File: tests/conftest.py
Comment (Avantol13) [inline tests/conftest.py:10]: dead code

**Review by Avantol13 — CHANGES_REQUESTED** [review]

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:162]: I assume these values (especially `max_time` comes from some testing? It would nice if this script accepted that value as an argument but it's not necessary if you're certain this good after some tests. If this were code we expected to stick around, I'd argue for more config but I think it's okay

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:137]: How are you handling errors in this new code? It looks like they don't re-raise from this function, so it just continues to run new batches. Is that intended? Are people expected to parse the logs for errors and rerun with a start guid? If so, how do they specify where to "end", e.g. when things started working again. 

The alternative would be for the script to actually die when there's an error by re-raising the exception and providing output so the user knows what batch failed and what the last successful GUID was

**Review by Avantol13 — CHANGES_REQUESTED** [review]

File: bin/migrate_to_single_table.py
Comment (Avantol13) [inline bin/migrate_to_single_table.py:141]: The service isn't fully aware of the job, so I'd leave the specific details out. I'd just say "Please re-run the job with the following `start-did`: {last_seen_guid}"

Comment (Avantol13) [reply bin/migrate_to_single_table.py:141]: shouldn't this error handling also be around the bulk_insert_records?

**Review by Avantol13 — DISMISSED** [review]

File: bin/migrate_to_single_table.py
Comment (Avantol13) [reply bin/migrate_to_single_table.py:297]: Yeah but these aren't catching the exception just because the fields are [], right? Like urls can be empty, but `get_urls_record` should still return []. I'm talking about the case where there really is an exception and you catch it.

The db query fails b/c we lose connection to the db. So right now you log that, but you don't return anything or reraise the exception, so the code continues on. I don't _think_ we want that behavior. If the db connection is broken, we want to raise an exception. And higher up, if we get an exception, we want to retry getting the record a few times just in case it was something flaky.

File: bin/migrate_to_single_table.py
Comment (Avantol13) [reply bin/migrate_to_single_table.py:353]: I'd say we remove this then, it's probably adding a decent chunk of time iterating through all the records

File: pyproject.toml
Comment (Avantol13) [inline pyproject.toml:12]: need a version bump above

**Comment by Avantol13** [comment]
At this point, most of my major feedback has been addressed. I'm dismissing my blocking review requesting changes so that progress can be made if I don't get to this before I head out on PTO.

Things I was not able to complete:

- Manual testing, running the service the old way, migrating, then running the new way

Things that need to be finished before a final approval:

- Passing integration tests
- Ideally another dev manually testing

File: indexd/single_table_settings.py
Comment (paulineribeyre) [inline indexd/single_table_settings.py:3]: Most of `single_table_settings` is a copy/merge of `default_settings` and `default_test_settings` - I notice a few changes but not sure why they're needed, for example `DIST` is different and `DRS_SERVICE_INFO.id` is not in all the settings files

Is adding a new file necessary? Can't we switch between `SQLAlchemyIndexDriver` and `SingleTableSQLAlchemyIndexDriver` depending on another config field, or even an env var?

If that's not possible, can we remove the duplicated settings and just overwrite `default_settings` with whatever is in `single_table_settings`?
Maybe `single_table_settings` could do `from indexd.default_settings import *` like `default_test_settings` does, and just set a different `CONFIG["INDEX"]`

File: pyproject.toml
Comment (paulineribeyre) [inline pyproject.toml:15]: you mentioned this can be removed now

File: indexd/index/drivers/single_table_alchemy.py
Comment (paulineribeyre) [inline indexd/index/drivers/single_table_alchemy.py:215]: there was a comment in the multi-table driver that i think we should keep:
```
if page is not None:
    # order by updated date so newly added stuff is
    # at the end (reduce risk that a new records ends up in a page
    # earlier on) and allows for some logic to check for newly added records
    # (e.g. parallelly processing from beginning -> middle and ending -> middle
    #       and as a final step, checking the "ending"+1 to see if there are
    #       new records).
    query = query.order_by(IndexRecord.updated_date)
else:
    query = query.order_by(IndexRecord.did)
```

File: indexd/index/drivers/single_table_alchemy.py
Comment (paulineribeyre) [inline indexd/index/drivers/single_table_alchemy.py:985]: in the multi-table code, these dates were validated by `_validate_and_format_content_dates`, do we not need to validate here anymore?

File: indexd/index/drivers/single_table_alchemy.py
Comment (paulineribeyre) [inline indexd/index/drivers/single_table_alchemy.py:571]:
```suggestion
                        + " Falling back to 'file_upload' on '/data_file'."
```

File: indexd/index/drivers/single_table_alchemy.py
Comment (paulineribeyre) [inline indexd/index/drivers/single_table_alchemy.py:729]:
```suggestion
            # delete this GUID's aliases
```

File: indexd/index/drivers/single_table_alchemy.py
Comment (paulineribeyre) [inline indexd/index/drivers/single_table_alchemy.py:1507]:
```suggestion
    for url in url_metadata:
```

File: indexd/index/drivers/single_table_alchemy.py
Comment (paulineribeyre) [inline indexd/index/drivers/single_table_alchemy.py:991]: i don't think `check_url_metadata()` needs to return anything, there are no changes to the input `urls_metadata` - the only benefit i see is being able to write this one-liner 😄

File: indexd/index/drivers/single_table_alchemy.py
Comment (paulineribeyre) [inline indexd/index/drivers/single_table_alchemy.py:1048]:
```suggestion
            new_record.rev = str(uuid.uuid4())[:8]
```

File: indexd/index/drivers/single_table_alchemy.py
Comment (paulineribeyre) [inline indexd/index/drivers/single_table_alchemy.py:1522]: does something break if you remove these?

File: migrations/versions/bb3d7586a096_createsingletable.py
Comment (paulineribeyre) [reply migrations/versions/bb3d7586a096_createsingletable.py:20]: Can we add a TODO in this file?

File: tests/conftest.py
Comment (paulineribeyre) [inline tests/conftest.py:9]: i think that was the last use of it, so we should be able to remove the `cdisutilstest` dependency! 🎉

File: deployment/Secrets/indexd_settings.py
Comment (BinamB) [reply deployment/Secrets/indexd_settings.py:6]: We change the `CONFIG["INDEX"]["driver"]` from `SQLAlchemyIndexDriver` to `SingleTableSQLAlchemyIndexDriver`
Should i do it a different way? Some logic to make it simple?

File: tests/conftest.py
Comment (BinamB) [reply tests/conftest.py:71]: There are tests that shouldn't use the fixture

File: .secrets.baseline
Comment (paulineribeyre) [reply .secrets.baseline:239]: Can you mention the migration in "Deployment changes", say it's optional and how to run it

File: deployment/Secrets/indexd_settings.py
Comment (paulineribeyre) [reply deployment/Secrets/indexd_settings.py:6]: i think a config flag would nice, something like
```
# one of: "multitable" (default), "single_table"
DATABASE_FORMAT = ""
```
and then initializing `CONFIG["INDEX"]` with
```
if DATABASE_FORMAT.lower() == "single_table":
    CONFIG["INDEX"] = ...
else:
    CONFIG["INDEX"] = ...
```
but it depends on what we want to do about [this](https://github.com/uc-cdis/indexd/pull/376#discussion_r1747361996)

File: tests/conftest.py
Comment (paulineribeyre) [reply tests/conftest.py:71]: for tests that shouldn't use the fixture, shouldn't we use the single table by default instead of the multitable? [here](https://github.com/uc-cdis/indexd/blob/bad7c84aa036d2471bb9be99a5c049521be595ab/tests/conftest.py#L110-L115)
since single table will eventually be the default

File: tests/conftest.py
Comment (paulineribeyre) [inline tests/conftest.py:101]: we might want to log the exception instead of silently passing (same in `app` and `user`)

File: tests/conftest.py
Comment (paulineribeyre) [inline tests/conftest.py:139]: the `add` and `delete` functions still exist in the auth driver, why does this need to change?

File: tests/test_driver_alchemy_crud.py
Comment (paulineribeyre) [inline tests/test_driver_alchemy_crud.py:32]: i'm confused why you write `driver = SQLAlchemyIndexDriver(POSTGRES_CONNECTION)` in a bunch of tests that use `combined_default_and_single_table_settings`. When we use the single table it shouldn't be that driver(?)

there are 2 `test_driver_init_does_not_create_records` tests, the other one doesn't include `driver = SQLAlchemyIndexDriver(POSTGRES_CONNECTION)`, is it needed here?

File: tests/test_driver_alchemy_crud.py
Comment (paulineribeyre) [inline tests/test_driver_alchemy_crud.py:894]: "Deployment changes" should mention that sqlite databases are not supported anymore

File: tests/test_migration.py
Comment (paulineribeyre) [inline tests/test_migration.py:1]: could we rename this file `test_migrate_to_single_table`? and if possible, move it to `postgres/migrations`?

File: indexd/index/drivers/single_table_alchemy.py
Comment (BinamB) [reply indexd/index/drivers/single_table_alchemy.py:991]: Changed it. Running check_url_metadata first and then setting the url_metadata outside of the function

File: tests/conftest.py
Comment (BinamB) [reply tests/conftest.py:139]: This ones a little complicated. Its the way tests are set. I did it a while ago and idk why it doesnt work but when i remove the driver, the tests start to fail

File: tests/test_migration.py
Comment (BinamB) [reply tests/test_migration.py:1]: I changed the name of the file but moving it to the `postgres/migrations` directory makes the second test get stuck at migration for some reason

File: tests/test_driver_alchemy_crud.py
Comment (BinamB) [reply tests/test_driver_alchemy_crud.py:32]: Went ahead and cleaned this up. I removed the `combined_default_and_single_table_settings` since these are tests that test direct queries to the old tables

File: indexd/index/drivers/single_table_alchemy.py
Comment (BinamB) [reply indexd/index/drivers/single_table_alchemy.py:1522]: removed

File: indexd/single_table_settings.py
Comment (BinamB) [reply indexd/single_table_settings.py:3]: Removing the new file and consolidating everything in conftest

File: indexd/index/drivers/single_table_alchemy.py
Comment (paulineribeyre) [inline indexd/index/drivers/single_table_alchemy.py:1019]: `_validate_and_format_content_dates` should probably be called before `session.add` and `session.commit`

File: docs/migration_to_single_table_indexd.md
Comment (BinamB) [reply docs/migration_to_single_table_indexd.md:17]: Yep! I worked with PE to come of with this

File: indexd/index/drivers/single_table_alchemy.py
Comment (BinamB) [reply indexd/index/drivers/single_table_alchemy.py:1019]: oops

File: indexd/index/drivers/single_table_alchemy.py
Comment (paulineribeyre) [inline indexd/index/drivers/single_table_alchemy.py:1010]:
```suggestion
```
it looks like `_validate_and_format_content_dates` sets the right values, so we shouldn't overwrite them

**Review by paulineribeyre — APPROVED** [review]
🎉 🚀

**Pull request merged**
